File: src/main/java/org/tindalos/principle/domain/checker/DesignQualityDetectorsRunner.java
package org.tindalos.principle.domain.checker;

import java.util.List;

import org.tindalos.principle.domain.core.DesignQualityCheckConfiguration;
import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.core.logging.TheLogger;
import org.tindalos.principle.domain.coredetector.CheckInput;
import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.coredetector.Detector;

import com.google.common.collect.Lists;

public class DesignQualityDetectorsRunner {

	private final List<Detector> detectors;

	public DesignQualityDetectorsRunner(List<Detector> detectors) {
		this.detectors = Lists.newArrayList(detectors);
	}

	public DesignQualityDetectorsRunner(Detector... detectors) {
		this.detectors = Lists.newArrayList(detectors);
	}

	public DesignQualityCheckResults execute(List<Package> packages, DesignQualityCheckConfiguration designQualityCheckConfiguration) {

		List<CheckResult> checkResults = Lists.newArrayList();
		CheckInput checkInput = new CheckInput(packages, designQualityCheckConfiguration);
		for (Detector detector : detectors) {
			if (detector.isWanted(designQualityCheckConfiguration.getExpectations())) {
			    TheLogger.info(detector + " is running");
				CheckResult checkResult = detector.analyze(checkInput);
				checkResults.add(checkResult);
			}
		}
		return new DesignQualityCheckResults(checkResults);
	}

}

File: src/main/java/org/tindalos/principle/domain/checker/DesignQualityDetectorsRunner.scala
package org.tindalos.principle.domain.checker

import org.tindalos.principle.domain.core.logging.TheLogger
import org.tindalos.principle.domain.core.{DesignQualityCheckConfiguration, Package}
import org.tindalos.principle.domain.coredetector.{CheckInput, Detector}

object DesignQualityDetectorsRunner {

  def buildDetectorsRunner(detectors: List[Detector]) =

    (packages: List[Package], designQualityCheckConfiguration: DesignQualityCheckConfiguration) => {

      val checkInput = new CheckInput(packages, designQualityCheckConfiguration)

      val checkResults = for (detector <- detectors
                              if detector.isWanted(designQualityCheckConfiguration.expectations))
                          yield runDetector(checkInput, detector)

      new DesignQualityCheckResults(checkResults.flatten)
    }

  private def runDetector(checkInput: CheckInput, detector: Detector) =
    try {
      TheLogger.info(detector + " is running.")
      Some(detector.analyze(checkInput))
    } catch {
      case unwantedException: RuntimeException => TheLogger.error(unwantedException.getMessage)
        None
    }

}